ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+
+  semian_cb_data_unlock(self);
+  return retval;
+
+}
+
+static VALUE
+semian_cb_data_array_set_index(VALUE self, VALUE idx, VALUE val)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  // check shared memory for NULL
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (TYPE(idx) != T_FIXNUM && TYPE(idx) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+  if (TYPE(val) != T_FIXNUM && TYPE(val) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  int index = NUM2INT(idx);
+  double value = NUM2DBL(val);
+
+  if (index <0 || index >= ptr->arr_max_size) {
+    return Qnil;
+  }
+
+  if (!semian_cb_data_lock(self)){
+    return Qnil;
+  }
+
+  ptr->shm_address->errors[index] = value;
+  ptr->shm_address->arr_length = index+1;
+
+  semian_cb_data_unlock(self);
+  return val;
+
+}
+
+static VALUE
+semian_cb_data_array_length(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+  int arr_length =ptr->shm_address->arr_length;
+  semian_cb_data_unlock(self);
+  return INT2NUM(arr_length);
+}
+
+static VALUE
+semian_cb_data_set_push_back(VALUE self, VALUE num)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+  if (TYPE(num) != T_FIXNUM && TYPE(num) != T_FLOAT/*|| TYPE(size) != T_BIGNUM*/)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  shared_cb_data *data = ptr->shm_address;
+  if (data->arr_length == ptr->arr_max_size) {
+    int i;
+    for (i=1; i< ptr->arr_max_size; ++i){
+      data->errors[i-1] = data->errors[i];
+    }
+    --(data->arr_length);
+  }
+  data->errors[(data->arr_length)] = NUM2DBL(num);
+  ++(data->arr_length);
+  semian_cb_data_unlock(self);
+  return self;
+}
+
+static VALUE
+semian_cb_data_set_pop_back(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  shared_cb_data *data = ptr->shm_address;
+  if (0 == data->arr_length)
+    retval = Qnil;
+  else {
+    retval = DBL2NUM(data->errors[data->arr_length-1]);
+    --(data->arr_length);
+  }
+
+  semian_cb_data_unlock(self);
+  return retval;
+}
+
+static VALUE
+semian_cb_data_set_pop_front(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  shared_cb_data *data = ptr->shm_address;
+  if (0 >= data->arr_length)
+    retval = Qnil;
+  else {
+    retval = DBL2NUM(data->errors[0]);
+    int i=0;
```

Comment:
Any reason you're not doing this decl. in the `for`? AFAIK in C99 you declare variables at start of function, in modern C you declare them inline. I don't really know which one is common in the Rubby world @csfrancis would know